### ISSUE TITLE
Support Managed Identity

### ISSUE THREAD

**Issue by DarqueWarrior** [Azure/static-web-apps#88]
Are there plans to support managed identity for static web apps?

I wanted to use managed identity with static web apps like I do with websites or functions. I stored information in Key Valut and could not auth my static web app using a managed identity. I had to resort to using a service principal.

**Comment by anganti** [comment]
@DarqueWarrior Could you elaborate a little more when you say used service principal. Was this on the client layer or the functions layer. I would assume you don't want to resolve secrets on the client side. Just trying to understand the context and scenarios where you would want to use Managed Identity besides functions.

**Comment by DarqueWarrior** [comment]
Sure! Yes, it was on the function side. I stored my CosmosDB connection string in Key Vault because I could not figure out how to create an input binding for CosmosDB in static web apps.

So after storing my connection string in Key Vault I wanted to auth my static web app with Key Vault. Without an managed identity I just used a service principal in the function to auth to Key Vault and grab my connection string.

**Comment by rajyraman** [comment]
I am having the same issue. On a standalone Azure Function, I can use managed identity to connect to CDS, but I am unclear on how to set this up in SWA api.

**Comment by rajyraman** [comment]
I ended up using EnvironmentCredential with client secret.

**Comment by tcc-sejohnson** [comment]
I'll readily add my vote to this -- it would be nice to support Managed Identity on the Azure Functions side, as I'd prefer to authorize the managed identity access to my Azure SQL DB than have to create a login and provide the creds through a connection string.

**Comment by ARambazamba** [comment]
Secret less config using managed identity would be a great feature

**Comment by demo4life** [comment]
Maybe you can vote here: https://feedback.azure.com/forums/34192--general-feedback/suggestions/43528497-pricipal-id-static-web-app

**Comment by anthonychu** [comment]
Thanks. Currently our guidance is to use the [bring your own function app](https://docs.microsoft.com/en-us/azure/static-web-apps/functions-bring-your-own) feature if you need managed identity in your function app. We do want to bring this to Static Web Apps’ built-in managed functions at some point but do not have an ETA.

**Comment by turdwaster** [comment]
@anthonychu Any update on the plan for this? Not being able to access a key vault from a managed function despite the SWA having an identity is really unproductive.

**Comment by hansmbakker** [comment]
This is not only about Key Vault - it is also about authenticating Azure Functions against a CosmosDB instance without needing secret management for example.

@anthonychu - this feature would really make working with the Static Web App's Managed Functions much nicer - at the moment it feels like the Managed Functions are an overly restricted feature and one would almost wonder what its usecase is when bring-your-own Functions are available. Since Functions are an integration component they will need a connection to other components in most of the cases and having managed identity support would really make that much better.

**Comment by anthonychu** [comment]
\+ @Reshmi-Sriram

**Comment by hansmbakker** [comment]
> This is not only about Key Vault - it is also about authenticating Azure Functions against a CosmosDB instance without needing secret management for example.
> This feature would really make working with the Static Web App's Managed Functions much nicer - at the moment it feels like the Managed Functions are a restricted feature and it's not clear what usecase it's meant for when bring-your-own Functions are available?
> Since Functions are an integration component they will need a connection to other components in most of the cases and having managed identity support would really make that much better.

@Reshmi-Sriram can you answer this?

**Comment by Reshmi-Sriram** [comment]
Hey @hansmbakker,
Thank you for the inputs! We have taken them into consideration and the engineering team has triaged this feature ask.
We do not have a solid timeline yet, but expect to hear back from us regarding the status in another 2-3 months. Thanks!

**Comment by hansmbakker** [comment]
@Reshmi-Sriram thank you, looking forward to your update!

**Comment by hansmbakker** [comment]
@Reshmi-Sriram I saw the [SWA database connections](https://learn.microsoft.com/en-us/azure/static-web-apps/database-overview) and wanted to understand whether that is the direction you were looking at to cover this feature request, or that you are looking something else?

I cannot understand from the documentation whether the SWA database connection feature is meant to be used from the Managed Functions, or that it is meant to be used only from the frontend code?

If it is meant to be used only from the frontend code, then I'm afraid the SWA database connection feature does not cover the feature request of this issue (using a Managed Identity in general, from Managed Functions)

**Comment by Reshmi-Sriram** [comment]
Hi @hansmbakker,
You're right. This is a parallel feature we have worked on, and in no way means to close this ticket. 
Like I mentioned, the engineering team is currently triaging the support for Managed Identity in Managed Functions, and we expect to give updates only in 2-3 months timeline. 
Please wait to hear more from us, but we're definitely looking into this!
cc// @thomasgauvin @mkarmark

**Comment by hansmbakker** [comment]
Hi @Reshmi-Sriram, thank you for clarifying this!

And the database connection feature is definitely a very useful feature for quickly scaffolding an API layer over people's databases!

**Comment by foundify** [comment]
Hello, any updates regarding managed identity with managed functions in SWA?

**Comment by hansmbakker** [comment]
@Reshmi-Sriram is the update you planned to give available somewhere?

**Comment by thomasgauvin** [comment]
Hey folks, providing an update on this thread, managed identity support is planned for end of 2024, it has been delayed due to dependencies on other teams that support the managed identity service.

**Comment by Chizaruu** [comment]
I will wait patiently. 

No Managed Functions identity support sucks when trying to access multiple databases, especially when using SvelteKit.  (No BYO Functions for me :c)

**Comment by jrblanno** [comment]
I would be amazing to use Managed identities in SWA :(

**Comment by hansmbakker** [comment]
The last update we got was

> [managed identity support is planned for end of 2024](https://github.com/Azure/static-web-apps/issues/88#issuecomment-1904195235)

I see @thomasgauvin switched jobs. It is now almost Q2 2025 and I'm not sure whether Microsoft staff is looking at this repo much anymore?

@Timothyw0 it seems you were the only Microsoft-representative who commented for the last few months - could you speak to the status of this issue?

**Comment by chuanqisun** [comment]
Is it possible for the team to look into this again? Our tenant (MSFT) has disabled secret-based authentication for Entra ID apps. Without managed identity, we would lose the benefit of having an all-in-one platform to host our apps. We are considering leaving the platform and switching to old school Web App Service. We really hope to stay with Static Web App as it simplifies the local dev experience. Please help us keep using Azure Static Web App ❤️

**Comment by LuKePicci** [comment]
@chuanqisun try with bring-your-own-function app first, you will have managed identities there, no need to switch into old school App Service. You have to upgrade the SKU of SWA instance, but it will still cost less then a full App Service and SWA+Function would still be much more constraining (for the good) in terms of architectural design for static server-side web assets.

**Comment by chudamanisanju** [comment]
Authentication Error with Managed Identity in Azure Static Web Apps
Hi team,

We're running into an issue using @azure/identity with ManagedIdentityCredential in an Azure Static Web App backend API. Here's the error we're receiving:

`error: {"name":"CredentialUnavailableError"}
message: "ManagedIdentityCredential: Authentication failed. Message undefined: Error(s): Not Available - Timestamp: Not Available - Description: Unable to load the proper Managed Identity. - Correlation ID: xxxxx-022d-xxxx-90e8-xxxxxxxxx - Trace ID: Not Available"`



Context:
We're deploying a Next.js app with Azure Static Web Apps, and also using its integrated API functionality (Oryx-built Azure Functions).

The backend API attempts to use ManagedIdentityCredential to authenticate with Azure services (e.g., Key Vault).

Everything builds and deploys successfully, but runtime API calls fail due to the above Managed Identity error.

🧠 Hypothesis:
From research and testing, it seems Azure Static Web Apps does not support Managed Identity in the integrated API environment, possibly due to sandbox restrictions.

❓ Questions:
Can Managed Identity be used at all in the context of Static Web App backend APIs?

If not, is there a recommended secure alternative (e.g., using App Registration credentials via environment variables)?

Are there any plans to add support for Managed Identity in Azure Static Web Apps backend environments?

Any guidance or recommendations would be greatly appreciated.

Thanks!

**Comment by LuKePicci** [comment]
> Managed Identity be used at all in the context of Static Web App backend APIs?

@chudamanisanju yes but you have to use bring-your-own-function for the SWA backend. If you opt for Azure Function App as bring-your-own-function backend there would be no much differences in terms of development and production environment. This also implies modifying your deployment workflows (e.g. azure devops pipelines) in order to skip the API backend building in SWA step, and then building and deploying it in a separate step using the usual Function App tasks.
 
> If not, is there a recommended secure alternative (e.g., using App Registration credentials via environment variables)?

IMHO it is better to switch your SWA into bring-your-own-function mode than moving your app security model away from managed identities.

**Comment by DigitalFlow** [comment]
Voting for this too, would be great to have.
